crates/oxc_transformer/src/es2017/async_to_generator.rs

```diff
@@ -65,138 +74,318 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+            _ => None,
+        };
+
+        if let Some(new_expr) = new_expr {
+            *expr = new_expr;
         }
     }
 
     fn exit_statement(&mut self, stmt: &mut Statement<'a>, ctx: &mut TraverseCtx<'a>) {
-        if let Statement::FunctionDeclaration(func) = stmt {
-            if !func.r#async || func.generator {
-                return;
+        let new_statement = match stmt {
+            Statement::FunctionDeclaration(func) => self.transform_function_declaration(func, ctx),
+            Statement::ExportDefaultDeclaration(decl) => {
+                if let ExportDefaultDeclarationKind::FunctionDeclaration(func) =
+                    &mut decl.declaration
+                {
+                    self.transform_function_declaration(func, ctx)
+                } else {
+                    None
+                }
             }
-            let new_function = self.transform_function(func, ctx);
-            if let Some(id) = func.id.take() {
-                *stmt = ctx.ast.statement_declaration(ctx.ast.declaration_variable(
-                    SPAN,
-                    VariableDeclarationKind::Const,
-                    ctx.ast.vec1(ctx.ast.variable_declarator(
-                        SPAN,
-                        VariableDeclarationKind::Const,
-                        ctx.ast.binding_pattern(
-                            ctx.ast.binding_pattern_kind_from_binding_identifier(id),
-                            NONE,
-                            false,
-                        ),
-                        Some(ctx.ast.expression_from_function(new_function)),
-                        false,
-                    )),
-                    false,
-                ));
-            } else {
-                *stmt =
-                    ctx.ast.statement_declaration(ctx.ast.declaration_from_function(new_function));
+            Statement::ExportNamedDeclaration(decl) => {
+                if let Some(Declaration::FunctionDeclaration(func)) = &mut decl.declaration {
+                    self.transform_function_declaration(func, ctx)
+                } else {
+                    None
+                }
             }
+            _ => None,
+        };
+
+        if let Some(new_statement) = new_statement {
+            self.ctx.statement_injector.insert_after(stmt.address(), new_statement);
         }
     }
+
+    fn exit_method_definition(
+        &mut self,
+        node: &mut MethodDefinition<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        self.transform_function_for_method_definition(&mut node.value, ctx);
+    }
 }
 
 impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
-    fn transform_function(
+    /// Transforms `await` expressions to `yield` expressions.
+    /// Ignores top-level await expressions.
+    #[allow(clippy::unused_self)]
+    fn transform_await_expression(
         &self,
-        func: &mut Function<'a>,
+        expr: &mut Box<'a, AwaitExpression<'a>>,
```

Comment:
`&mut Box<'a, AwaitExpression<'a>>` is a double-reference (`&mut` is a reference, and so is `Box`). It might make more sense to take a `&mut AwaitExpression<'a>`.